keps/sig-network/2438-dual-stack-apiserver/README.md

```diff
@@ -230,80 +234,177 @@ scenario of wrong-single-stack pods...
+
+          - It reconciles the Service if and only if it knows what the
+            Service's secondary IP should be (that is, if and only if
+            it has a dual-stack `--service-cluster-ip-range`). If it
+            only has a single service CIDR configured then it just
+            ignores the Service, assuming that other apiservers are
+            reconciling it correctly, and that its own configuration
+            will presumably be updated to match theirs soon.
+
+      - If the `DualStackAPIServer` feature gate is not enabled on
+        this apiserver, then it ignores both the secondary
+        EndpointSlice and the Service, assuming that other apiservers
+        are reconciling them correctly, and that its own configuration
+        will presumably be updated to match theirs soon.
+
+      - (If there are IPs of the secondary address family in etcd, but
+        there are not actually any running dual-stack apiservers, then
+        the leases on those IPs will eventually expire, and then all
+        remaining apiservers will agree that the Service needs to be
+        forced back to single-stack.)
+
+(The above description assumes the `lease` reconciliation mode; we
+could support `master-count` mode as well, by using the `NodeName`
+fields in Endpoints and EndpointSlice to correlate secondary IPs with
+primary ones to keep the two sets in sync. But alternatively, we could
+just say that dual-stack apiservers are only compatible with `lease`
+mode.)
 
-(Note that the middle case covers both "upgrading from
-`DualStackAPIServer=false` to `DualStackAPIServer=true`" _and_
-"upgrading the cluster from single stack to dual stack (after
-`DualStackAPIServer` goes GA)". In both cases, some apiservers will
-begin advertising dual-stack endpoints before other apiservers are
-aware of the configuration change, but the un-upgraded apiservers will
-allow the upgrade to proceed smoothly.)
+```
+<<[UNRESOLVED you-get-an-endpointslice-YOU-get-an-endpointslice-EVERYONE-GETS-AN-ENDPOINTSLICE ]>>
+
+Tim had suggested in the original KEP PR:
+
+> I think this "does not reconcile" is a bug and when we eventually
+> fix that, it will flap. I was thinking about alternatives...
+>
+> EPSlice already says that a given IP might be in more than one slice
+> (e.g. while rebalancing). What if EACH apiserver wrote their OWN
+> epslices. Single-stack servers would write 1, dual-stack would write
+> 2. Cleanup on downgrade is tricky - which slices are "mine" ? Maybe
+> we need some other way to ident apiserver replicas so they can
+> operate independently in this manner?
+
+While this is not necessary, given that I think the solution I
+described above should work fine, it might be interesting to do
+something along these lines anyway, since it would then guarantee that
+every cluster has at least one multi-slice Service in it, thus
+ensuring that every EndpointSlice client would be forced to deal
+properly with multi-slice Services...
```

Comment:
@robscott ...

Reply:
I like this. As far as cleanup/ownership, do we have any concept of node name or any other unique identifier per apiserver?

Reply:
The apiservers already have a mechanism for cleaning up stale IPs. (*Two* mechanisms actually; this PR adds a bunch of background explanation about that.) We'd just modify it so if an apiserver decided an IP was stale, it would delete the entire slice containing that IP rather than just removing the IP from the `Addresses` field of a singular slice.

Reply:
I guess Tim's suggestion might have been "each apiserver writes out a slice containing all of the apiserver IPs it knows about". I was thinking "each apiserver writes out a slice containing only its own IP". (Or "pair of slices" for dual-stack.) I think that's cleaner.

(But we could make the cleanup work either way; in the "all IPs" version, when an apiserver reconciled, it would write out its own slice with all the IPs it thinks are live, and then delete any slice that contained an IP it considered stale.)

Reply:
Yeah, I was thinking of each apiserver writing 1 or 2 slices, each holding its own IP(s).  So a stable HA cluster would have 3  slices for each family.

If we used the same lease approach it could clean itself up.  

I'm not sure if it is any better than the approach above, and it's certainly a bit more complex.  So may not be worth doing.